src/bigint.rs

```diff
@@ -3254,6 +3254,70 @@ impl BigInt {
     pub fn trailing_zeros(&self) -> Option<u64> {
         self.data.trailing_zeros()
     }
+
+    /// Returns whether the bit in position `bit` is set,
+    /// uses the two's complement for negative numbers
+    pub fn bit(&self, bit: u64) -> bool {
+        // Let the binary representation of a number be
+        //   ... 0  x 1 0 ... 0
+        // Then the two's complement is
+        //   ... 1 !x 1 0 ... 0
+        // where !x is obtained from x by flipping each bit
+        if bit >= u64::from(big_digit::BITS) * self.len() as u64 {
+            self.is_negative()
+        } else if self.is_negative() && bit > self.data.trailing_zeros().unwrap() {
+            !self.data.bit(bit)
+        } else {
+            self.data.bit(bit)
+        }
+    }
+
+    /// Sets or clears the bit in the given position,
+    /// uses the two's complement for negative numbers
+    pub fn set_bit(&mut self, bit: u64, value: bool) {
+        match self.sign {
+            Sign::Plus => self.data.set_bit(bit, value),
+            Sign::NoSign => {
+                // clearing a bit for zero is a no-op
+                if value {
+                    self.data.set_bit(bit, true);
+                    self.sign = Sign::Plus;
+                }
+            }
+            Sign::Minus => {
+                let bits_per_digit = u64::from(big_digit::BITS);
+                // The first part of this `if` condition is not necessary but included because
+                // computing trailing_zeros can be avoided when the bit to set/clear is outside
+                // the represented digits
+                if bit >= bits_per_digit * self.len() as u64
+                    || bit > self.data.trailing_zeros().unwrap()
+                {
+                    self.data.set_bit(bit, !value);
+                } else {
+                    // This is the general case that basically corresponds to what `bitor_neg_pos`
+                    // (when setting bit) or `bitand_neg_pos` (when clearing bit) does, except there
+                    // is no need to explicitly iterate over the digits of the right-hand side
```

Comment:
I have to admit, that bit-twiddling has fallen out of my memory by now. Your adaptation looks appropriate, but I can't help feeling that there must be a way to shorten the scan both before and after the target bit. That is, before only needs to figure out the carry state, and after might be able to stop on zero carry. I'm not sure.

Maybe this is a good candidate for some fuzzing? There are tests of this sort in `ci/big_quickcheck`.

Reply:
You are absolutely correct, in most cases it will iterate over far more than necessary. When writing the code, I was in doubt of whether I should go for more efficient and complicated code, or sub-optimal code that was fairly straightforward and shorter. I went for the latter, but I sort of regret that now. I am currently rewriting the `set_bit` method, so stay tuned!